File: backend-express/src/services/auth.service.ts
import { AuthResponse } from '../types';
import { RegisterRequest, LoginRequest, RefreshRequest } from '../types';
import { userService } from './user.service';
import { jwtService } from './jwt.service';
import { DuplicateEmailError } from '../errors/DuplicateEmailError';
import { UnauthorizedError } from '../errors/UnauthorizedError';
import { NotFoundError } from '../errors/NotFoundError';

export class AuthService {
  async register(request: RegisterRequest): Promise<AuthResponse> {
    if (await userService.existsByEmail(request.email)) {
      throw new DuplicateEmailError('Email already exists');
    }

    const user = await userService.createUser(request.email, request.password);
    const accessToken = jwtService.generateAccessToken(user.email);
    const refreshToken = jwtService.generateRefreshToken(user.email);

    return {
      accessToken,
      refreshToken,
      email: user.email,
    };
  }

  async login(request: LoginRequest): Promise<AuthResponse> {
    const user = await userService.findByEmail(request.email);
    
    if (!user) {
      throw new UnauthorizedError('Invalid email or password');
    }

    const isPasswordValid = await userService.verifyPassword(
      request.password,
      user.passwordHash
    );

    if (!isPasswordValid) {
      throw new UnauthorizedError('Invalid email or password');
    }

    const accessToken = jwtService.generateAccessToken(user.email);
    const refreshToken = jwtService.generateRefreshToken(user.email);

    return {
      accessToken,
      refreshToken,
      email: user.email,
    };
  }

  async refresh(request: RefreshRequest): Promise<AuthResponse> {
    if (!jwtService.validateToken(request.refreshToken)) {
      throw new UnauthorizedError('Invalid or expired refresh token');
    }

    const email = jwtService.getEmailFromToken(request.refreshToken);
    const user = await userService.findByEmail(email);

    if (!user) {
      throw new NotFoundError('User not found');
    }

    const newAccessToken = jwtService.generateAccessToken(user.email);
    const newRefreshToken = jwtService.generateRefreshToken(user.email);

    return {
      accessToken: newAccessToken,
      refreshToken: newRefreshToken,
      email: user.email,
    };
  }
}

export const authService = new AuthService();

File: backend-spring/src/main/java/com/verdego/api/auth/AuthService.java
package com.verdego.api.auth;

import com.verdego.api.common.exception.DuplicateEmailException;
import com.verdego.api.security.JwtService;
import com.verdego.api.user.Role;
import com.verdego.api.user.User;
import com.verdego.api.user.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional
public class AuthService {

    private final UserService userService;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    public AuthResponse register(RegisterRequest request) {
        if (userService.existsByEmail(request.getEmail())) {
            throw new DuplicateEmailException("Email already exists");
        }

        User user = User.builder()
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .roles(List.of(Role.USER))
                .build();

        user = userService.createUser(user);
        String accessToken = jwtService.generateAccessToken(user.getEmail());
        String refreshToken = jwtService.generateRefreshToken(user.getEmail());

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .email(user.getEmail())
                .build();
    }

    public AuthResponse login(LoginRequest request) {
        User user = userService.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Invalid email or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new RuntimeException("Invalid email or password");
        }

        String accessToken = jwtService.generateAccessToken(user.getEmail());
        String refreshToken = jwtService.generateRefreshToken(user.getEmail());

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .email(user.getEmail())
                .build();
    }

    public AuthResponse refresh(RefreshRequest request) {
        String refreshToken = request.getRefreshToken();
        
        if (!jwtService.validateToken(refreshToken)) {
            throw new RuntimeException("Invalid or expired refresh token");
        }

        String email = jwtService.getEmailFromToken(refreshToken);
        User user = userService.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));

        String newAccessToken = jwtService.generateAccessToken(user.getEmail());
        String newRefreshToken = jwtService.generateRefreshToken(user.getEmail());

        return AuthResponse.builder()
                .accessToken(newAccessToken)
                .refreshToken(newRefreshToken)
                .email(user.getEmail())
                .build();
    }
}
